File: src/dstream/compressing_algo.zig
const std = @import("std");

const aux = @import("../_auxlib.zig");

/// Does this algorithm have the capacity to ingest a data item at logical time
/// T?
///
/// @param u Unsigned integer type for operands and return value.
/// @param S The number of buffer sites available.
/// @param T Queried logical time.
/// @returns Whether there is capacity to ingest at time T.
pub fn has_ingest_capacity(comptime u: type, S: u, T: u) bool {
    aux.assert_unsigned(u);
    _ = T;
    return S > 0;
}

/// Site selection for compressing curation with even buffer size.
///
/// Site 0 is special (always holds T=0). Remaining S-1 sites are managed
/// with modulus M = S - 1.
fn assign_storage_site_even_S(comptime u: type, S: u, T: u) u {
    std.debug.assert(S & 1 == 0);

    if (T == 0) return 0;

    const M = S - 1;
    const T_ = T - 1;
    const si = aux.bit_length(u, T_ / M); // Current sampling interval
    const h: u = @ctz(@max(T_, 1)); // Hanoi value
    if (h < si) // discard without storing
        return S
    else
        return T_ % M + 1;
}

/// Site selection for compressing curation with odd buffer size.
///
/// No special site. All S sites participate uniformly with modulus M = S.
fn assign_storage_site_odd_S(comptime u: type, S: u, T: u) u {
    std.debug.assert(S & 1 == 1);

    const M = S;
    const si = aux.bit_length(u, T / M); // Current sampling interval
    const h: u = @ctz(@max(T, 1)); // Hanoi value
    if (h < si) // discard without storing
        return S
    else
        return T % M;
}

/// Site selection for compressing curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @param u Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be greater than 0.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns S if no site should be selected (i.e., discard).
pub fn assign_storage_site(comptime u: type, S: u, T: u) u {
    aux.assert_unsigned(u);
    std.debug.assert(has_ingest_capacity(u, S, T));

    return if (S & 1 == 0)
        assign_storage_site_even_S(u, S, T)
    else
        assign_storage_site_odd_S(u, S, T);
}

File: src/dstream/compressing_algo.rs
use crate::_auxlib as aux;

/// Does this algorithm have the capacity to ingest a data item at logical time
/// T?
///
/// @template Uint Unsigned integer type for operands.
/// @param S The number of buffer sites available.
/// @param T Queried logical time.
/// @returns Whether there is capacity to ingest at time T.
#[allow(non_snake_case)]
pub fn has_ingest_capacity<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> bool {
    _ = T;
    S > Uint::zero()
}

/// Site selection for compressing curation with even buffer size.
///
/// Site 0 is special (always holds T=0). Remaining S-1 sites are managed
/// with modulus M = S - 1.
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size. Must be positive and even.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns S if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
fn _assign_storage_site_even_s<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
    let _0: Uint = Uint::zero();
    let _1: Uint = Uint::one();

    if T == _0 {
        return _0;
    }

    let m: Uint = S - _1;
    let t_: Uint = T - _1;
    let si: Uint = aux::bit_length::<Uint>(t_ / m); // Current sampling interval
    let h: Uint = aux::ctz::<Uint>(std::cmp::max(t_, _1)); // Hanoi value
    if h < si {
        S // discard without storing
    } else {
        t_ % m + _1
    }
}

/// Site selection for compressing curation with odd buffer size.
///
/// No special site. All S sites participate uniformly with modulus M = S.
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size. Must be positive and odd.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns S if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
fn _assign_storage_site_odd_s<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
    let _1: Uint = Uint::one();

    let m: Uint = S;
    let si: Uint = aux::bit_length::<Uint>(T / m); // Current sampling interval
    let h: Uint = aux::ctz::<Uint>(std::cmp::max(T, _1)); // Hanoi value
    if h < si {
        S // discard without storing
    } else {
        T % m
    }
}

/// Site selection implementation for compressing curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be greater than 0.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns S if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
pub fn _assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
    debug_assert!(has_ingest_capacity(S, T));
    let _1: Uint = Uint::one();

    if S & _1 == Uint::zero() {
        _assign_storage_site_even_s::<Uint>(S, T)
    } else {
        _assign_storage_site_odd_s::<Uint>(S, T)
    }
}

/// Site selection implementation for compressing curation.
///
/// What buffer site should the T'th data item be stored to?
///
/// @template Uint Unsigned integer type for operands and return value.
/// @param S Buffer size.
///     Must be greater than 0.
/// @param T Current logical time.
/// @returns The selected storage site, if any.
///     Returns None if no site should be selected (i.e., discard).
#[allow(non_snake_case)]
pub fn assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Option<Uint> {
    let k = _assign_storage_site(S, T);
    if k == S {
        None
    } else {
        Some(k)
    }
}

pub struct Algo;

#[allow(non_snake_case)]
impl crate::dstream::HasIngestCapacityTrait for Algo {
    fn has_ingest_capacity<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> bool {
        has_ingest_capacity::<Uint>(S, T)
    }
}

#[allow(non_snake_case)]
impl crate::dstream::AssignStorageSiteTrait for Algo {
    fn _assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Uint {
        _assign_storage_site::<Uint>(S, T)
    }

    fn assign_storage_site<Uint: aux::UnsignedTrait>(S: Uint, T: Uint) -> Option<Uint> {
        assign_storage_site::<Uint>(S, T)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_smoke_has_ingest_capacity() {
        has_ingest_capacity::<u32>(16, 101);
    }

    #[test]
    fn test_smoke_impl_assign_storage_site() {
        _assign_storage_site::<u32>(16, 101);
    }

    #[test]
    fn test_smoke_assign_storage_site() {
        assign_storage_site::<u32>(16, 101);
    }
}
